--- ipchains-1.3.10/ipchains.c.qswitch	Fri Sep  1 08:33:04 2000
+++ ipchains-1.3.10/ipchains.c	Mon Sep 24 09:38:44 2001
@@ -785,7 +785,7 @@
 }
 
 int
-main(int argc, char *argv[])
+real_main(int argc, char *argv[])
 {
 	unsigned int nsaddrs = 0, ndaddrs = 0;
 	struct in_addr *saddrs = NULL, *daddrs = NULL;
@@ -808,6 +808,10 @@
 
 	program = argv[0];
 
+	/* Some reinitialization for when called multiple times */
+	memset(&fw, '\0', sizeof(fw));
+	optind = 0;
+
 	while ((c = getopt_long(argc, argv,
 	   "-A:C:D:R:I:L::F::Z::N:X::P:MS:Vh::o::p:s:d:j:i:fbvm:nlt:xy",
 					   opts, NULL)) != -1) {
@@ -1365,13 +1369,13 @@
 		int sockfd;
 		if ((sockfd = socket(AF_INET, SOCK_RAW, IPPROTO_RAW)) == -1) {
 			perror("ipchains: socket creation failed");
-			exit(1);
+			return 1;
 		}
 
 		if (setsockopt(sockfd, IPPROTO_IP, IP_FW_MASQ_TIMEOUTS,
 			       (char *) &timeouts, sizeof(timeouts)) != 0) {
 			perror("ipchains: setting MASQ timeouts failed");
-			exit(1);
+			return 1;
 		}
 		close(sockfd);
 		ret = 1;
@@ -1387,7 +1391,62 @@
 
 	if (!ret)
 		fprintf(stderr, "ipchains: %s\n", ipfwc_strerror(errno));
-	exit(!ret);
+	return !ret;
+}
+
+int
+main(int argc, char *argv[])
+{
+	static char buf[BUFSIZ];
+	int rc = 0;
+
+	program = argv[0];
+
+	if (argc >= 2 && (!strcmp(argv[1], "-Q") || !strcmp(argv[1], "--quick")))
+	{
+		if (argc != 3)
+			exit_error(PARAMETER_PROBLEM, "-Q option requires filename or ``-''");
+
+		if (strcmp(argv[2], "-"))
+			if (freopen(argv[2], "r", stdin) == NULL)
+				exit_error(PARAMETER_PROBLEM, "Failed to open '%s'", argv[2]);
+
+		while (fgets(buf, BUFSIZ, stdin))
+		{
+			static char *sep = " \t\n\r";
+			char *av[100];
+			char *tok;
+			int ac = 0;
+
+			for (tok = strtok(buf, sep); ac < 100 && tok; tok = strtok(NULL, sep))
+			{
+				char *comment = strchr(tok, '#'); /* e.g. #!/sbin/ipchains -Q */
+
+				if (comment)
+					*comment = '\0';
+
+				if (tok != comment)
+					av[ac++] = tok;
+
+				if (comment)
+					break;
+			}
+
+			if (ac == 100)
+				exit_error(PARAMETER_PROBLEM, "Too many parameters: %s", buf);
+
+			av[ac] = NULL;
+
+			if (ac)
+				rc += real_main(ac, av);
+		}
+	}
+	else
+	{
+		rc += real_main(argc, argv);
+	}
+
+	return rc;
 }
 
 static int check_inverse(const char option[], int *invert)
@@ -2536,9 +2595,10 @@
 "       %s -[LFZNX] [chain] [options]\n"
 "       %s -P chain target [options]\n"
 "       %s -M [ -L | -S ] [options]\n"
+"       %s -Q cmdfile\n"
 "       %s -h [icmp] (print this help information, or ICMP list)\n\n",
 	       package_version, program, program, program, program, program,
-	       program, program);
+	       program, program, program);
 
 	printf(
 "Commands:\n"
@@ -2564,7 +2624,9 @@
 "  --masquerading\n"
 "            -M -L		List current masquerading connections\n"
 "  --set     -M -S tcp tcpfin udp\n"
-"				Set masquerading timeout values\n\n"
+"				Set masquerading timeout values\n"
+"  --quick    -Q cmdfile	Read multiple ipchains commands from cmdfile\n"
+"				(read from standard input if cmdfile is ``-'')\n\n"
 
 "Options:\n"
 "  --bidirectional -b		insert two rules: one with -s & -d reversed\n"
